types/react/index.d.ts

```diff
@@ -78,7 +79,12 @@ declare namespace React {
     type ComponentType<P = {}> = ComponentClass<P> | StatelessComponent<P>;
 
     type Key = string | number;
-    type Ref<T> = string | { bivarianceHack(instance: T | null): any }["bivarianceHack"];
+
+    interface RefObject<T> {
+        readonly current: T | null;
```

Comment:
My guess is it will be more annoying than useful to treat `current` as nullable (even though it technically is).

Reply:
This is the same behavior as the current argument to ref callbacks; you also can brute force your way through it by using a `!`.

Reply:
Yes, but in 2.7+ you can use definite assignment assertion to pretend the ref is non-null:

```ts
class Foo extends React.Component {
  elem!: HTMLInputElement

  render() {
    return <input ref={elem => elem && (this.elem = elem)} />;
  }

  componentDidMount() {
    this.elem.focus();
  }
}
```

No such recourse will exist with `current` if it's nullable. Also, if people want to opt in to a nullable ref they can do so:

```ts
class Foo extends React.Component {
  ref = React.createRef<HTMLInputElement | null>();

  render() {
    return <input ref={this.ref} />;
  }
}
```

But if it's required to be nullable in the definition, they are unable to opt out.

Reply:
The definite assignment is a different problem. You are checking if `elem` is `null` or not with `elem && (this.elem = elem)`, and so you cannot assign the `null` value to it. The definite assignment lets you avoid the `undefined` state, but `null` is a normal part of the API.

Reply:
What it allows you to do is treat `this.elem` as non-null everywhere _else, i.e. to avoid having to use `this.elem!` everywhere.

Reply:
That behavior is only correct if your subtree is always mounted. If your subtree is mounted conditionally, you now have a hidden `null` reference. The `| null` here is _correct behavior_.

Reply:
Yes, and if you happen to know it’s always mounted this is exactly the use case for the definite assignment assertion feature. It’s a small compromise in type safety but very convenient for situations where you know better than the compiler. This is why the feature exists. By forcing `current` to be `| null` we lose the option to exercise such discretion, so it’s a step backwards. Again, you can always opt in to perfect type safety by declaring the type parameter of the `RefObject` to be `| null`, but you cannot opt out if the decision is made for you.

Reply:
No, this is _not_ the use case for definite assignment assertion. The definite assignment is for avoiding the presence of `undefined` _specifically_.

Do not confuse `null` with `undefined`.

Reply:
Hm, you may be right, but how is it a potential memory leak? Doesn't React release instances once unmounted, thereby releasing their fields?